scrapy/core/scheduler.py

```diff
@@ -1,42 +1,169 @@
+    """
+    def __instancecheck__(cls, instance):
+        return cls.__subclasscheck__(type(instance))
+
+    def __subclasscheck__(cls, subclass):
+        return (
+            hasattr(subclass, "has_pending_requests") and callable(subclass.has_pending_requests)
+            and hasattr(subclass, "enqueue_request") and callable(subclass.enqueue_request)
+            and hasattr(subclass, "next_request") and callable(subclass.next_request)
+        )
+
+
+class BaseScheduler(metaclass=BaseSchedulerMeta):
+    """
+    The scheduler component is responsible for storing requests received from
+    the engine, and feeding them back upon request (also to the engine).
+
+    The original sources of said requests are:
+    * Spider: ``start_requests`` method, requests created for URLs in the ``start_urls`` attribute, request callbacks
+    * Spider middleware: ``process_spider_output`` and ``process_spider_exception`` methods
+    * Downloader middleware: ``process_request``, ``process_response`` and ``process_exception`` methods
+
+    The order in which the scheduler returns its stored requests (via the ``next_request`` method)
+    plays a great part in determining the order in which those requests are downloaded.
+
+    The methods defined in this class constitute the minimal interface that the Scrapy engine will interact with.
     """
-    def __init__(self, dupefilter, jobdir=None, dqclass=None, mqclass=None,
-                 logunser=False, stats=None, pqclass=None, crawler=None):
+
+    @classmethod
+    def from_crawler(cls, crawler: Crawler):
+        """
+        Factory method which receives the current :class:`~scrapy.crawler.Crawler` object as argument.
+        """
+        return cls()
+
+    def open(self, spider: Spider) -> Optional[Deferred]:
+        """
+        Called when the spider is opened by the engine. It receives the spider
+        instance as argument and it's useful to execute initialization code.
+
+        :param spider: the spider object for the current crawl
+        :type spider: :class:`~scrapy.spiders.Spider`
+        """
+        pass
+
+    def close(self, reason: str) -> Optional[Deferred]:
+        """
+        Called when the spider is closed by the engine. It receives the reason why the crawl
+        finished as argument and it's useful to execute cleaning code.
+
+        :param reason: a string which describes the reason why the spider was closed
+        :type reason: :class:`str`
+        """
+        pass
+
+    @abstractmethod
+    def has_pending_requests(self) -> bool:
+        """
+        ``True`` if the scheduler has enqueued requests, ``False`` otherwise
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def enqueue_request(self, request: Request) -> bool:
+        """
+        Process a request received by the engine.
+
+        Return ``True`` if the request is to be considered scheduled, ``False`` otherwise.
```

Comment:
I think it would be best to extend this to make it more clear why you would ever return `False` (e.g. your scheduler stores requests in an external service that is temporarily unreachable?), and what’s the consequence of doing so (is the request discarded, force-feeded into the engine, re-enqueued later?).